libyul/Utilities.cpp

```diff
@@ -86,56 +86,113 @@ std::string solidity::yul::reindent(std::string const& _code)
 	return out.str();
 }
 
-u256 solidity::yul::valueOfNumberLiteral(Literal const& _literal)
+LiteralValue solidity::yul::valueOfNumberLiteral(std::string_view const _literal)
 {
-	yulAssert(_literal.kind == LiteralKind::Number, "Expected number literal!");
+	return LiteralValue{LiteralValue::Data(_literal), std::string(_literal)};
+}
 
-	static std::map<YulString, u256> numberCache;
-	static YulStringRepository::ResetCallback callback{[&] { numberCache.clear(); }};
+LiteralValue solidity::yul::valueOfStringLiteral(std::string_view const _literal)
+{
+	std::string const s(_literal);
+	return LiteralValue{u256(h256(s, h256::FromBinary, h256::AlignLeft)), s};
+}
 
-	auto&& [it, isNew] = numberCache.try_emplace(_literal.value, 0);
-	if (isNew)
-	{
-		std::string const& literalString = _literal.value.str();
-		yulAssert(isValidDecimal(literalString) || isValidHex(literalString), "Invalid number literal!");
-		it->second = u256(literalString);
-	}
-	return it->second;
+LiteralValue solidity::yul::valueOfBuiltinStringLiteralArgument(std::string_view _literal)
+{
+	return LiteralValue{std::string(_literal)};
 }
 
-u256 solidity::yul::valueOfStringLiteral(Literal const& _literal)
+LiteralValue solidity::yul::valueOfBoolLiteral(std::string_view const _literal)
 {
-	yulAssert(_literal.kind == LiteralKind::String, "Expected string literal!");
-	yulAssert(_literal.value.str().size() <= 32, "Literal string too long!");
+	if (_literal == "true")
+		return LiteralValue{true};
+	else if (_literal == "false")
+		return LiteralValue{false};
 
-	return u256(h256(_literal.value.str(), h256::FromBinary, h256::AlignLeft));
+	yulAssert(false, "Unexpected bool literal value!");
 }
 
-u256 yul::valueOfBoolLiteral(Literal const& _literal)
+LiteralValue solidity::yul::valueOfLiteral(std::string_view const _literal, LiteralKind const& _kind, bool const _unlimitedLiteralArgument)
 {
-	yulAssert(_literal.kind == LiteralKind::Boolean, "Expected bool literal!");
+	switch (_kind)
+	{
+	case LiteralKind::Number:
+		return valueOfNumberLiteral(_literal);
+	case LiteralKind::Boolean:
+		return valueOfBoolLiteral(_literal);
+	case LiteralKind::String:
+		return _unlimitedLiteralArgument ? valueOfBuiltinStringLiteralArgument(_literal) : valueOfStringLiteral(_literal);
+	}
+	util::unreachable();
+}
+
+std::string solidity::yul::formatLiteral(solidity::yul::Literal const& _literal, bool const assertValidity)
```

Comment:
```suggestion
std::string solidity::yul::formatLiteral(solidity::yul::Literal const& _literal, bool const _assertValidity)
```

Maybe we could call it something like `_assumeValidity` though? The fact that we enforce the assumption with an assert is just an internal detail.

Reply:
Also, for invalid literals:

- Currently we'll return `"false"` for values other than `0` or `1`. If we have to choose one of the two, `"true"` would be more correct. But perhaps it would be best to print the numerical value instead?
- The final `return` will kick in also when the value of the hint does not match the numerical value. Its comment should mention that as one of the possibilities.
- What happens when you pass in an invalid number to `valueOfLiteral()`? Won't the conversion to `u256` throw or something?
    - Assuming in doesn't (it would probably show up in tests if it did), the result can't be the right number so the comparison is meaningless in that case. Perhaps we should not check that and just print the hint when we don't require the value to be valid? Now that I think of it, this bit should really be a part of validation, not formatting, shouldn't it?

Reply:
I see you moved the check that the hint matches the value to `validStringLiteral()`, but I think we should have it in `validNumberLiteral()` as well.

Reply:
Looks like other points from my comment above have been dealt with.